### PR TITLE
Contracts/tests/deposit and sell

### PR DESCRIPTION
**Pull request by W3stside**
### Deposit And Sell function testing

### CODE REVIEW

File: test/dutchExchange-DepositAndSell.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    } = contracts)
+
+    await Promise.all(testingAccs.map(acc => Promise.all([
+      eth.deposit({ from: acc, value: ETHBalance }),
+      eth.approve(dx.address, ETHBalance, { from: acc }),
+    ])))
+
+    await Promise.all(testingAccs.map(acc => Promise.all([
+      dx.deposit(eth.address, ETHBalance / 2, { from: acc }),
+    ])))
```
Comment (Velenir) [inline test/dutchExchange-DepositAndSell.js]: Can do without the second `Promise.all` here.

File: test/dutchExchange-DepositAndSell.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  })
+
+  it('Adds Token Pair', async () => dx.addTokenPair(eth.address, gno.address, 0, 0, 2, 1, { from: accounts[1] }))
```
Comment (Velenir) [inline test/dutchExchange-DepositAndSell.js]: unnecessary `async`

File: test/dutchExchange-DepositAndSell.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  it('DX Auction idx = 1 + Auction Start Time is > timestamp NOW [auction not started]', () => Promise.all(testingAccs.map(async (acc) => {
+    assert.equal(await getAuctionIndex(), 1, 'Auction index should be moved to 1')
+    assert.isAbove(await dx.getAuctionStart.call(eth.address, gno.address), timestamp(), 'auction time should be above now')
+  })))
+
+  it('DX balances cannot be more than amt deposited initially', () => Promise.all(testingAccs.map(async (acc) => {
+    assert(await dx.balances.call(eth.address, acc) <= ETHBalance / 2, 'Balances cannot be more than ETHBalance / 2')
```
Comment (Velenir) [inline test/dutchExchange-DepositAndSell.js]: Can use `assert.isAtMost` here, but whatever your preference.

File: test/dutchExchange-Tulip.js [hunk 4]
```diff
@@ -662,16 +667,427 @@ const c2 = () => contract('DutchExchange --> Tulip Flow --> 1 Seller + 2 Buyers
...
+    startingETH,
+    sellingAmount,
+    // startingGNO,
+    // ethUSDPrice,
+  } = startBal
+
+  before(async () => {
+    // get contracts
+    await setupContracts()
+    eventWatcher(dx, 'LogNumber', {})
```
Comment (Velenir) [inline test/dutchExchange-Tulip.js:695]: Actually you don't need the last argument, it's `{}` by default.

File: test/dutchExchange-Tulip.js [hunk 5]
```diff
@@ -662,16 +667,427 @@ const c2 = () => contract('DutchExchange --> Tulip Flow --> 1 Seller + 2 Buyers
...
+    CLAIMED FUNDS => ${claimedFunds.toEth()}
+    TULIPS ISSUED => ${tulipsIssued.toEth()}
+    `)
+
+    assert.equal(tulipsIssued, 0, 'Tulips only issued / minted after auction Close so here = 0')
+  })
+
+  it(
+    'BUYER1: Tries to lock and unlock Tulips --> Auction NOT cleared --> asserts 0 Tulips minted and in mapping', 
+    async () => unlockTulipTokens(buyer1),
```
Comment (Velenir) [inline test/dutchExchange-Tulip.js:815]: unnecessary `async`, favorite keyword?

File: test/dutchExchange-Tulip.js [hunk 6]
```diff
@@ -192,6 +194,11 @@ const c1 = () => contract('DutchExchange --> Tulip Flow --> 1 Seller + 1 Buyer |
     assert.equal(tulipsIssued, 0, 'Tulips only issued / minted after auction Close so here = 0')
   })
 
+  it(
+    'BUYER1: Tries to lock and unlock Tulips --> Auction NOT cleared --> asserts 0 Tulips minted and in mapping', 
+    async () => unlockTulipTokens(buyer1),
```
Comment (Velenir) [inline test/dutchExchange-Tulip.js:199]: unnecessary `async`

File: test/dutchExchange-Tulip.js [hunk 7]
```diff
@@ -662,16 +667,427 @@ const c2 = () => contract('DutchExchange --> Tulip Flow --> 1 Seller + 2 Buyers
...
 } else if (argv.c === 2) {
-  // fire contract 2
   c2()
+} else if (argv.c === 3) {
+  c3()
+} else if (argv.c === 4) {
+  c4()
 } else {
-  return Promise.all([c1(), c2()])
+  return Promise.all([c1(), c2(), c3(), c4()])
```
Comment (Velenir) [inline test/dutchExchange-Tulip.js:1092]: Are you sure we need `Promise.all` here. Don't `contract()` just schedule tests and are run sequentially.

File: test/dutchExchange-Tulip.js (hunk 5 above)
Comment (W3stside) [reply test/dutchExchange-Tulip.js:815]: no, just tired and not thinking properly, also yes

- will fix these in #160

File: test/dutchExchange-Tulip.js (hunk 7 above)
Comment (W3stside) [reply test/dutchExchange-Tulip.js:1092]: Haven't tested this, will check in #160

File: test/dutchExchange-DepositAndSell.js (hunk 1 above)
Comment (W3stside) [reply test/dutchExchange-DepositAndSell.js]: yep, refactored this and didn't catch it. thx

File: test/dutchExchange-DepositAndSell.js (hunk 2 above)
Comment (W3stside) [reply test/dutchExchange-DepositAndSell.js]: woops. what I get for 330am coding. thx

File: test/dutchExchange-Tulip.js (hunk 7 above)
Comment (W3stside) [reply test/dutchExchange-Tulip.js:1092]: yeah `forEach` is fine

**Review by Velenir — APPROVED** [review]

**Pull request merged**